### PR TITLE
Add job factory to enable custom job creation

### PR DESCRIPTION
**Pull request by ChrisPRobinson**
Job factory enables the ability to create multiple jobs from one base
job. It allows a job's initialize to take custom parameters in fed from
the factory.

1) Updated the manager to add a job factory with the associated job when
needed and added tests

2) Updated the cli job run and scheduler to ensure they can run and
schedule correctly

### CODE REVIEW

**Comment by ChrisPRobinson** [comment]
Tested with a sample set of job/factories like this
```
require 'etl'
require 'etl/job/base'
require 'etl/job_factory/base'
require 'singleton'
require 'etl/schedule/base'

module Test
  class MyTestJobFactory < ETL::JobFactory::Base
    def create(job_id, batch)
      klass = ::ETL::Job::Manager.instance.get_class(job_id)
      klass.new(batch, job_id,"1")
    end
  end
  
  class SpecialJob < ETL::Job::Base
    attr_accessor :thing1
    def self.register_job
      job_factory = ::Test::MyTestJobFactory.new
      ::ETL::Job::Manager.instance.register("special_job1", self, job_factory)
      ::ETL::Job::Manager.instance.register("special_job2", self, job_factory)
    end
    register_job
    
    def schedule
      return ::ETL::Schedule::Once.new(self, @batch) 
    end

    def initialize(batch, job_id, thing1)
       super(batch)
       @thing1 = thing1
       @id = job_id
    end
    
    def id
      @id
    end
    
    def run
      puts thing1
      super
    end
  end
end

class SpecialJob3 < ETL::Job::Base
  register_job
  
  def initialize(batch)
     super(batch)
  end
end
```

File: spec/job/manager_spec.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    end
+    def id
+      @id
+    end
+  end
+
+  class TestJob2 < ETL::Job::Base
+    def initialize(b)
+      super(b)
+    end
```
Comment (emya) [inline spec/job/manager_spec.rb]: I think we can remove this initialize

File: spec/job/manager_spec.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  # via the factory
+  class TestJob < ETL::Job::Base
+    attr_accessor :p1
+    def initialize(batch, id, p1)
+      super(batch)
+      @p1 = p1
+      @id = id
+    end
+    def id
+      @id
```
Comment (emya) [inline spec/job/manager_spec.rb]: attr_reader?

File: lib/etl/job/manager.rb [hunk 3]
```diff
@@ -3,31 +3,46 @@
...
+
     # Returns the job class registered for the specified ID, or nil if none
     def get_class(id)
       @job_classes[id]
     end
-    
-    # Iterates through the loaded classes specified during class initailiation.
+
+    # Returns the job class factory registered for the specified ID, or nil if none
+    def get_class_factory(job_id)
```
Comment (emya) [inline lib/etl/job/manager.rb]: Should we use `id` to keep consistency?

File: spec/job/manager_spec.rb (hunk 2 above)
Comment (ChrisPRobinson) [inline spec/job/manager_spec.rb]: This is ok

File: lib/etl/job/manager.rb (hunk 3 above)
Comment (ChrisPRobinson) [inline lib/etl/job/manager.rb]: I used job_id because the id could refer to the factory not the job

Comment (ChrisPRobinson) [inline lib/etl/job/manager.rb]: I wanted to make sure that it was known to be a job id

File: spec/job/manager_spec.rb (hunk 1 above)
Comment (ChrisPRobinson) [inline spec/job/manager_spec.rb]: removed

**Pull request merged**